src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs

```diff
@@ -121,38 +126,64 @@ protected override void FormatInlineCount(IList<string> commandClauses)
         {
             if (string.IsNullOrEmpty(path))
             {
                 var resultColumns = columns.Where(x => !HasMultipleSegments(x)).ToList();
                 if (excludePaths != null)
-                    resultColumns.AddRange(columns.Where(x => HasMultipleSegments(x) &&
-                        !excludePaths.Any(y => FormatFirstSegment(y).Contains(FormatFirstSegment(x)))));
+                {
+                    if (excludePaths.Count == 1 && excludePaths[0] == "*")
+                    {
+                        resultColumns.AddRange(columns.Where(x => HasMultipleSegments(x) &&
+                           !_session.Metadata.GetNavigationPropertyNames(collection.Name).Any(y => FormatFirstSegment(y).Contains(FormatFirstSegment(x)))));
+                    }
+                    else
+                    {
+                        resultColumns.AddRange(columns.Where(x => HasMultipleSegments(x) &&
+                           !excludePaths.Any(y => FormatFirstSegment(y).Contains(FormatFirstSegment(x)))));
+                    }
+                }
+                   
                 return resultColumns;
             }
             else
             {
-                return columns.Where(x => HasMultipleSegments(x) && FormatFirstSegment(x) == FormatFirstSegment(path))
-                    .Select(x => FormatSkipSegments(x, 1)).ToList();
+                var resultColumns = FormatFirstSegment(path) == "*"
+                    ? columns.Where(x => HasMultipleSegments(x) && _session.Metadata.GetNavigationPropertyNames(collection.Name).Contains(FormatFirstSegment(x)))
+                    : columns.Where(x => HasMultipleSegments(x) && FormatFirstSegment(x) == FormatFirstSegment(path));
+                return resultColumns.Select(x => FormatSkipSegments(x, 1)).ToList();
             }
         }
 
         private IList<KeyValuePair<string, bool>> SelectPathSegmentColumns(
-            IList<KeyValuePair<string, bool>> columns, string path, IList<string> excludePaths = null)
+            IList<KeyValuePair<string, bool>> columns, string path, EntityCollection collection, IList<string> excludePaths = null)
         {
             if (string.IsNullOrEmpty(path))
             {
                 var resultColumns = columns.Where(x => !HasMultipleSegments(x)).ToList();
                 if (excludePaths != null)
-                    resultColumns.AddRange(columns.Where(x => HasMultipleSegments(x) &&
-                        !excludePaths.Any(y => FormatFirstSegment(y).Contains(FormatFirstSegment(x)))));
+                {
+                    if (excludePaths.Count == 1 && excludePaths[0] == "*")
+                    {
+                        resultColumns.AddRange(columns.Where(x => HasMultipleSegments(x) &&
+                           !_session.Metadata.GetNavigationPropertyNames(collection.Name).Any(y => FormatFirstSegment(y).Contains(FormatFirstSegment(x)))));
```

Comment:
This is going to enumerate `_session.Metadata.GetNavigationPropertyNames(collection.Names)` roughly `columns.Count` times. You should hoist out the enumeration and `FormatFirstSegment` result into a set.

Reply:
No it's not enumerating `columns.Count` times, it's enumerating only the columns where `HasMultipleSegments` is getting hit. :)

W/E it's still a legit improvement, thanks for your contribution. Please check the new commits: Hopefully you can resolve this conversation.

Reply:
Apologies, I'm referring to the worst case performance.